### DISCUSSION TITLE
Is it possible to share CircuitBreaker state across several pipelines?

### DISCUSSION THREAD

**Discussion by dcavinar365** [App-vNext/Polly#2264] [Q&A] [answered]
I have a few different pipelines with slightly different configurations. As an example, one might be for idempotent operations that can retry and a separate one for operations that should not retry. There might also be different settings for timeouts.

However, they all access the same underlying resource and I want to use a common circuit breaker so I applied the same CircuitBreakerStrategyOptions instance to all the pipelines. 

However, it seems that each pipeline has its own circuit breaker CircuitStateController so when one breaks the others are still active. 

Is there a way to share circuit breaker states across multiple pipelines?

**Comment by martincostello** [comment]
If you take a look at our documentation for [pipeline registries](https://www.pollydocs.org/pipelines/resilience-pipeline-registry.html) you should be able to see how you can build up a pipeline composed of references to shared strategies.

**Reply by dcavinar365** [reply]
Thank you for the quick response @martincostello but I am struggling to understand from the provided examples there how I can build two or more strategies using the same circuit breaker.

**Reply by dcavinar365** [reply, marked as answer]
I figured out a way that is working in all my tests. Hopefully I didn't miss something crucial! 
First I have to create a ResiliencePipeline that just holds the desired circuit breaker strategy:
```
var cbPipeline = new ResiliencePipelineBuilder()
    .AddCircuitBreaker(options)
    .Build();
```
Then I can add that to any pipeline that needs that same state using the AddPipeline() builder extension method:
```
var pipeline = new ResiliencePipelineBuilder()
    .AddPipeline(cbPipeline)
    .Build();
```

**Reply by peter-csala** [reply]
Just a side-note: If you don't have a reference to the pipeline via a variable, then you can use the `ResiliencePipelineProvider`  to fetch the registered pipeline

```cs
services.AddResiliencePipeline("cb-pipeline", (builder, _) => builder.AddCircuitBreaker(cbOptions));
```
```cs
var pipelineProvider = context.ServiceProvider.GetRequiredService<ResiliencePipelineProvider<string>>();
var pipeline = new ResiliencePipelineBuilder()
    .AddPipeline(pipelineProvider.GetPipeline("cb-pipeline"))
    .Build();
```
